File: apps/frontend/src/services/gitea.ts
import api from './api';

export interface GiteaUser {
  login: string;
  name: string;
  email: string;
  avatar: string;
}

export interface Repository {
  id: number;
  name: string;
  full_name: string;
  description: string;
  private: boolean;
  fork: boolean;
  html_url: string;
  clone_url: string;
  ssh_url: string;
  language: string;
  size: number;
  updated_at: string;
  owner: {
    id: number;
    login: string;
    full_name: string;
    email: string;
    avatar_url: string;
  };
}

export interface FileContent {
  name: string;
  path: string;
  sha: string;
  size: number;
  url: string;
  html_url: string;
  git_url: string;
  download_url: string;
  type: string; // 'file' | 'dir'
  content?: string;
  encoding?: string;
}

export interface GiteaStatus {
  authorized: boolean;
  gitea_user?: GiteaUser;
}

// Gitea相关API
export const giteaAPI = {
  // 获取授权URL
  getAuthURL: (state?: string) => 
    api.get<{ auth_url: string }>(`/gitea/auth-url${state ? `?state=${state}` : ''}`),
  
  // OAuth回调
  callback: (code: string, state?: string) => 
    api.post('/gitea/callback', { code, state }),
  
  // 获取授权状态
  getStatus: () => 
    api.get<GiteaStatus>('/gitea/status'),
  
  // 撤销授权
  revokeAuth: () => 
    api.delete('/gitea/auth'),
  
  // 获取仓库列表
  getRepositories: () => 
    api.get<Repository[]>('/gitea/repositories'),
  
  // 获取目录内容
  getDirectoryContents: (owner: string, repo: string, path?: string) => 
    api.get<FileContent[]>(`/gitea/repos/${owner}/${repo}/contents${path ? `?path=${encodeURIComponent(path)}` : ''}`),
  
  // 获取文件内容
  getFileContent: (owner: string, repo: string, path: string) => 
    api.get<FileContent>(`/gitea/repos/${owner}/${repo}/file?path=${encodeURIComponent(path)}`),
};

export default giteaAPI;

File: apps/backend/services/gitea.go
package services

import (
	"backend/models"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"

	"golang.org/x/oauth2"
)

type GiteaService struct {
	config   *oauth2.Config
	baseURL  string
	client   *http.Client
}

func NewGiteaService() *GiteaService {
	baseURL := getEnv("GITEA_BASE_URL", "https://gitea.com")

	config := &oauth2.Config{
		ClientID:     getEnv("GITEA_CLIENT_ID", ""),
		ClientSecret: getEnv("GITEA_CLIENT_SECRET", ""),
		RedirectURL:  getEnv("GITEA_REDIRECT_URL", "http://localhost:5173/oauth/callback"),
		Scopes:       []string{"read:repository", "read:user"},
		Endpoint: oauth2.Endpoint{
			AuthURL:  fmt.Sprintf("%s/login/oauth/authorize", baseURL),
			TokenURL: fmt.Sprintf("%s/login/oauth/access_token", baseURL),
		},
	}

	return &GiteaService{
		config:  config,
		baseURL: baseURL,
		client:  &http.Client{},
	}
}

func (g *GiteaService) GetAuthURL(state string) string {
	return g.config.AuthCodeURL(state)
}

func (g *GiteaService) ExchangeCode(code string) (*oauth2.Token, error) {
	return g.config.Exchange(context.Background(), code)
}

func (g *GiteaService) GetUser(token *oauth2.Token) (*models.GiteaToken, error) {
	req, err := http.NewRequest("GET", fmt.Sprintf("%s/api/v1/user", g.baseURL), nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token.AccessToken))

	resp, err := g.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to get user info: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var user struct {
		ID       int64  `json:"id"`
		Login    string `json:"login"`
		FullName string `json:"full_name"`
		Email    string `json:"email"`
		Avatar   string `json:"avatar_url"`
	}

	if err := json.Unmarshal(body, &user); err != nil {
		return nil, err
	}

	giteaToken := &models.GiteaToken{
		AccessToken:  token.AccessToken,
		RefreshToken: token.RefreshToken,
		TokenType:    token.TokenType,
		GiteaUserID:  user.ID,
		GiteaLogin:   user.Login,
		GiteaName:    user.FullName,
		GiteaEmail:   user.Email,
		GiteaAvatar:  user.Avatar,
	}

	if !token.Expiry.IsZero() {
		giteaToken.ExpiresAt = &token.Expiry
	}

	return giteaToken, nil
}

func (g *GiteaService) GetRepositories(token string) ([]models.Repository, error) {
	req, err := http.NewRequest("GET", fmt.Sprintf("%s/api/v1/user/repos", g.baseURL), nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))

	resp, err := g.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to get repositories: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var repos []models.Repository
	if err := json.Unmarshal(body, &repos); err != nil {
		return nil, err
	}

	return repos, nil
}

func (g *GiteaService) GetFileContent(token, owner, repo, path string) (*models.FileContent, error) {
	url := fmt.Sprintf("%s/api/v1/repos/%s/%s/contents/%s", g.baseURL, owner, repo, path)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))

	resp, err := g.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to get file content: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var content models.FileContent
	if err := json.Unmarshal(body, &content); err != nil {
		return nil, err
	}

	// 解码base64内容
	if content.Encoding == "base64" && content.Content != "" {
		decoded, err := base64.StdEncoding.DecodeString(content.Content)
		if err != nil {
			return nil, err
		}
		content.Content = string(decoded)
		content.Encoding = "utf-8"
	}

	return &content, nil
}

func (g *GiteaService) GetDirectoryContents(token, owner, repo, path string) ([]models.FileContent, error) {
	url := fmt.Sprintf("%s/api/v1/repos/%s/%s/contents/%s", g.baseURL, owner, repo, path)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))

	resp, err := g.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to get directory contents: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var contents []models.FileContent
	if err := json.Unmarshal(body, &contents); err != nil {
		return nil, err
	}

	return contents, nil
}

func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}
